// @quasar/app-vite/lib/quasar-config-file.js
import { basename, dirname, isAbsolute, join, relative } from 'node:path'
import { pathToFileURL } from 'node:url'
import { existsSync, readFileSync } from 'node:fs'
import fse from 'fs-extra'
import { merge } from 'webpack-merge'
import { rolldown, watch as rolldownWatch } from 'rolldown'
import { transformAssetUrls } from '@quasar/vite-plugin'
import { isCI } from 'ci-info'

import { error, fatal, log, tip, warn } from './utils/logger.js'
import { appFilesValidations } from './utils/app-files-validations.js'
import { getPackageMajorVersion } from './utils/get-package-major-version.js'
import { resolveExtension } from './utils/resolve-extension.js'
import { ensureElectronArgv } from './utils/ensure-argv.js'
import { quasarRolldownVueShimPlugin } from './plugins/rolldown.vue-shim.js'
import {
  quasarRolldownInjectReplacementsDefine,
  quasarRolldownInjectReplacementsPlugin
} from './plugins/rolldown.inject-replacements.js'

import { findClosestOpenPort, localHostList } from './utils/net.js'
import { BASELINE_WIDELY_AVAILABLE_TARGET_STRING } from './utils/build-targets.js'
import { encodeForDiff } from './utils/encode-for-diff.js'
import { debounce } from './utils/rate-limit.js'
import {
  defaultBackendAppEnvPrefix,
  defaultClientAppEnvPrefix,
  getAppEnv,
  getQuasarConfEnv
} from './utils/env.js'

const quasarModesList = [
  'bex',
  'capacitor',
  'cordova',
  'electron',
  'pwa',
  'spa',
  'ssr'
]

const urlRegex = /^http(s)?:\/\//i
const defaultPortMapping = {
  spa: 9000,
  ssr: 9100, // 9150 for SSR + PWA
  pwa: 9200,
  electron: 9300,
  cordova: 9400,
  capacitor: 9500,
  bex: 9600
}

const quasarComponentRE = /^(Q[A-Z]|q-)/
const quasarConfigBanner = `/* oxlint-disable */
/**
 * THIS FILE IS GENERATED AUTOMATICALLY.
 * 1. DO NOT edit this file directly as it won't do anything.
 * 2. EDIT the original quasar.config file INSTEAD.
 * 3. DO NOT git commit this file. It should be ignored.
 *
 * This file is still here because there was an error in
 * the original quasar.config file and this allows you to
 * investigate the Node.js stack error.
 *
 * After you fix the original file, this file will be
 * deleted automatically.
 **/
`

function escapeHTMLTagContent(str) {
  return str ? str.replaceAll(/[<>]/g, '') : ''
}
function escapeHTMLAttribute(str) {
  return str ? str.replaceAll('"', '') : ''
}

function formatPublicPath(publicPath) {
  if (!publicPath) return '/'

  if (!publicPath.endsWith('/')) {
    publicPath = `${publicPath}/`
  }

  if (urlRegex.test(publicPath)) return publicPath

  if (!publicPath.startsWith('/')) {
    publicPath = `/${publicPath}`
  }

  return publicPath
}

function formatRouterBase(publicPath) {
  if (!publicPath || !publicPath.startsWith('http')) {
    return publicPath
  }

  const match = publicPath.match(
    /^(https?:)\/\/(([^:/?#]*)(?::([0-9]+))?)([/]{0,1}[^?#]*)(\?[^#]*|)(#.*|)$/
  )
  return formatPublicPath(match[5] || '')
}

function parseAssetProperty(prefix) {
  return asset => {
    if (typeof asset === 'string') {
      return {
        path: asset[0] === '~' ? asset.slice(1) : `${prefix}/${asset}`
      }
    }

    return {
      ...asset,
      path:
        typeof asset.path === 'string'
          ? asset.path[0] === '~'
            ? asset.path.slice(1)
            : `${prefix}/${asset.path}`
          : asset.path
    }
  }
}

function getUniqueArray(original) {
  return [...new Set(original)]
}

function uniquePathFilter(value, index, self) {
  return self.map(obj => obj.path).indexOf(value.path) === index
}

const extRE = /\.[m|c]?[j|t]s$/
function formatQuasarAssetPath(asset, type) {
  return asset.includes('/')
    ? extRE.test(asset)
      ? asset
      : `${asset}.js`
    : `quasar/${type}/${asset}.js`
}

let cachedExternalHost,
  addressRunning = false

async function onAddress({ host, port }, mode) {
  if (
    ['cordova', 'capacitor'].includes(mode) &&
    (!host || localHostList.includes(host.toLowerCase()))
  ) {
    if (cachedExternalHost) {
      host = cachedExternalHost
    } else {
      const { getExternalIP } = await import('./utils/get-external-ip.js')
      host = await getExternalIP()
      cachedExternalHost = host
    }
  }

  try {
    const openPort = await findClosestOpenPort(port, host)
    if (port !== openPort) {
      warn()
      warn(`️️Setting port to closest one available: ${openPort}`)
      warn()

      port = openPort
    }
  } catch (err) {
    warn()

    if (err.message === 'ERROR_NETWORK_PORT_NOT_AVAIL') {
      warn(
        'Could not find an open port. Please configure a lower one to start searching with.'
      )
    } else if (err.message === 'ERROR_NETWORK_ADDRESS_NOT_AVAIL') {
      warn(
        'Invalid host specified. No network address matches. Please specify another one.'
      )
    } else {
      warn('Unknown network error occurred')
      console.error(err)
    }

    warn()

    if (!addressRunning) process.exit(1)
    return null
  }

  addressRunning = true
  return { host, port }
}

let wsToken = null
function createWsToken() {
  if (wsToken === null) {
    const list =
      'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
    const listLen = list.length
    wsToken = ''
    for (let i = 0; i < 12; i++) {
      wsToken += list.at(Math.floor(Math.random() * listLen))
    }
  }

  return wsToken
}

export class QuasarConfigFile {
  /** @type {import('../types/configuration/context').QuasarContext} */
  #ctx
  #opts
  #versions = {}
  #address
  #tempFile
  #hostPackageJsonPath
  #confEnv
  #appEnv = null
  #appEnvFilesChanged = true
  #shouldFail = true

  #watch = {
    buildId: 0,
    chokidarWatch: null,
    isBuilding: false,
    resolveRead: null,
    onUpdate: null,
    shouldReadBuiltFileAgain: false,
    rolldownWatcher: null,
    confEnvWatcher: null,
    appEnvWatcher: null,
    cachedQuasarConf: null
  }

  #cssVariables
  #storeProvider
  #vueDevtools
  #electronInspectPort

  constructor({ ctx, host, port, verifyAddress, watch = false }) {
    this.#ctx = ctx
    this.#opts = { host, port, verifyAddress, watch }

    this.#hostPackageJsonPath = ctx.appPaths.resolve.app('package.json')
    this.#confEnv = this.#getConfEnv()

    // if filename syntax gets changed, then also update the "clean" cmd
    this.#tempFile = `${ctx.appPaths.quasarConfigFilename}.temporary.compiled.${Date.now()}.js`
  }

  /**
   * @returns {import('rolldown').RolldownOptions}
   */
  #getRolldownConfig() {
    const { appPaths } = this.#ctx

    return {
      platform: 'node',

      input: appPaths.quasarConfigFilename,

      output: {
        file: this.#tempFile,
        format: 'esm',
        codeSplitting: false,
        banner: quasarConfigBanner
      },

      /**
       * Required for Windows host, otherwise Node.js will complain
       * about importing with absolute paths and will fail with
       * ESM error ("ERR_UNSUPPORTED_ESM_URL_SCHEME"):
       */
      makeAbsoluteExternalsRelative: true,
      external: /node_modules/,

      resolve: {
        // Define the aliases which have to be usable in the quasar.config file
        alias: {
          '#q-app': '@quasar/app-vite'
        },

        extensions: ['.js', '.ts', '.json']
      },

      transform: {
        target: 'node22',
        define: {
          ...this.#confEnv.envDefineList,
          ...quasarRolldownInjectReplacementsDefine,

          /**
           * Avoid "import.meta.env is not defined"
           * as this is not available in Node.js.
           *
           * Should be the last statement in "define",
           * otherwise it will override all previous statements.
           */
          'import.meta.env': '{}'
        }
      },

      plugins: [
        quasarRolldownInjectReplacementsPlugin(),
        quasarRolldownVueShimPlugin()
      ],

      // .quasar/tsconfig.json won't be available for the first time executing dev/build/prepare.
      // So, Rolldown will error out.
      // We need to suppress the error. Otherwise, it will be noisy and cause a temp file to be created.
      // tsconfig is not really important for the config file itself anyway.
      tsconfig: false,

      watch: {
        exclude: /node_modules/
      }
    }
  }

  watch(onUpdate) {
    this.#watch.onUpdate = debounce(onUpdate, 500)
    if (!this.#watch.isBuilding && this.#watch.shouldReadBuiltFileAgain) {
      this.#readBuiltFile().then(quasarConf => {
        this.#injectAppEnv(quasarConf)
        this.#watch.onUpdate(quasarConf)
      })
    }
  }

  async read() {
    const { appPaths, cacheProxy, appExt } = this.#ctx

    await Promise.all([
      cacheProxy.getModule('cssVariables').then(cssVariables => {
        this.#cssVariables = cssVariables
      }),
      cacheProxy.getModule('storeProvider').then(storeProvider => {
        this.#storeProvider = storeProvider
      })
    ])

    await appExt.registerAppExtensions()

    if (this.#ctx.mode.capacitor) {
      const { capVersion } = await cacheProxy.getModule('capCli')

      const getCapPluginVersion = name => {
        const version = getPackageMajorVersion(name, appPaths.capacitorDir)
        return version === void 0 ? false : version || true
      }

      Object.assign(this.#versions, {
        capacitor: capVersion,
        capacitorPluginApp: getCapPluginVersion('@capacitor/app'),
        capacitorPluginSplashscreen: getCapPluginVersion(
          '@capacitor/splash-screen'
        )
      })
    }

    const { promise, resolve } = this.#opts.watch
      ? Promise.withResolvers()
      : {
          promise: this.#build(),
          resolve: null
        }

    if (this.#opts.watch) {
      const { watch } = await import('chokidar')
      this.#watch.chokidarWatch = watch
      this.#watch.resolveRead = resolve
      this.#createConfEnvWatcher()
      this.#watchBuild()
    }

    return promise
      .then(() => this.#readBuiltFile())
      .then(quasarConf => {
        this.#injectAppEnv(quasarConf)
        this.#shouldFail = false
        return quasarConf
      })
  }

  async #readBuiltFile() {
    this.#watch.shouldReadBuiltFileAgain = false
    let quasarConfigFn

    try {
      // cache busting it, hence the ?t= param
      const fnResult = await import(
        pathToFileURL(this.#tempFile).href + '?t=' + Date.now()
      )
      quasarConfigFn = fnResult.default || fnResult
    } catch (err) {
      console.log()
      console.error(err)

      if (!this.#opts.watch) {
        fatal(
          'The quasar.config file has runtime errors. Please check the Node.js stack above against the' +
            ` temporarily created ${basename(this.#tempFile)} file, fix the original file` +
            ' then DELETE the temporary one ("quasar clean --qconf" can be used).',
          'FAIL'
        )
      }

      const msg =
        'Importing quasar.config file results in error. Please check the' +
        ` Node.js stack above against the temporarily created ${basename(this.#tempFile)} file` +
        ' and fix the original file then DELETE the temporary one ("quasar clean --qconf" can be used).'

      if (this.#shouldFail) fatal(msg, 'FAIL')

      warn(msg + '\n')
      return
    }

    const quasarConf = await this.#computeConfig(quasarConfigFn)

    if (this.#opts.watch && quasarConf !== void 0) {
      this.#watch.cachedQuasarConf = quasarConf
    }

    return quasarConf
  }

  #getConfEnv() {
    const { quasarCli } = JSON.parse(
      fse.readFileSync(this.#hostPackageJsonPath, 'utf8')
    )

    return getQuasarConfEnv({
      ctx: this.#ctx,
      envCfg: quasarCli?.quasarConfEnv || {},
      useSnapshot: this.#opts.watch
    })
  }

  #injectAppEnv(quasarConf) {
    if (
      this.#appEnvFilesChanged ||
      this.#appEnv?.snapshot.envCfg !== encodeForDiff(quasarConf.build.env)
    ) {
      this.#appEnvFilesChanged = false

      const newAppEnv = getAppEnv({
        ctx: this.#ctx,
        envCfg: quasarConf.build.env,
        useSnapshot: this.#opts.watch
      })

      if (
        !this.#opts.watch ||
        this.#appEnv?.snapshot.watchEnvFiles !==
          newAppEnv.snapshot?.watchEnvFiles ||
        this.#appEnv?.snapshot.envCfg !== newAppEnv.snapshot?.envCfg
      ) {
        log(newAppEnv.envBanner)
      }

      if (this.#opts.watch) {
        if (this.#watch.appEnvWatcher === null) {
          this.#createAppEnvWatcher(newAppEnv)
        } else if (
          newAppEnv.snapshot.watchEnvFiles !==
          this.#appEnv.snapshot.watchEnvFiles
        ) {
          const watcher = this.#watch.appEnvWatcher
          watcher.unwatch([
            ...this.#appEnv.watchEnvFiles.difference(newAppEnv.watchEnvFiles)
          ])
          watcher.add([
            ...newAppEnv.watchEnvFiles.difference(this.#appEnv.watchEnvFiles)
          ])
        }
      }

      this.#appEnv = newAppEnv
    }

    quasarConf.metaConf.clientEnvDefineList = this.#appEnv.clientEnvDefineList
    quasarConf.metaConf.backendEnvDefineList = this.#appEnv.backendEnvDefineList
  }

  #handleAppEnvChange = debounce(() => {
    if (
      this.#watch.cachedQuasarConf !== null &&
      !this.#watch.shouldReadBuiltFileAgain &&
      !this.#watch.isBuilding
    ) {
      this.#injectAppEnv(this.#watch.cachedQuasarConf)
      this.#watch.onUpdate?.(this.#watch.cachedQuasarConf)
    }
  }, 300)

  #createConfEnvWatcher() {
    const { appDir } = this.#ctx.appPaths
    const onEnvChange = debounce(changedFile => {
      const newConfEnv = this.#getConfEnv()

      if (
        newConfEnv.snapshot.watchEnvFiles !==
        this.#confEnv.snapshot.watchEnvFiles
      ) {
        const watcher = this.#watch.confEnvWatcher
        watcher.unwatch([
          ...this.#confEnv.watchEnvFiles.difference(newConfEnv.watchEnvFiles)
        ])
        watcher.add([
          ...newConfEnv.watchEnvFiles.difference(this.#confEnv.watchEnvFiles)
        ])
      }

      if (
        newConfEnv.snapshot.envDefineList !==
        this.#confEnv.snapshot.envDefineList
      ) {
        log()
        log(
          `Detected quasar.config env change from ${relative(appDir, changedFile)}`
        )

        this.#confEnv = newConfEnv

        if (this.#watch.rolldownWatcher !== null) {
          const watcher = this.#watch.rolldownWatcher
          this.#watch.rolldownWatcher = null
          watcher.close() // don't wait for it, we're in a hurry
        }

        // we're rebuilding everything anyway,
        // so avoid currently scheduled app env changes
        this.#handleAppEnvChange.cancel()

        this.#watchBuild()
        return
      }

      this.#confEnv = newConfEnv
    }, 300)

    this.#watch.confEnvWatcher = this.#watch
      .chokidarWatch(
        [this.#hostPackageJsonPath, ...this.#confEnv.watchEnvFiles],
        {
          ignoreInitial: true
        }
      )
      .on('add', onEnvChange)
      .on('change', onEnvChange)
      .on('unlink', onEnvChange)
  }

  #createAppEnvWatcher(initialAppEnv) {
    const onChange = () => {
      this.#appEnvFilesChanged = true
      this.#handleAppEnvChange()
    }
    this.#watch.appEnvWatcher = this.#watch
      .chokidarWatch([...initialAppEnv.watchEnvFiles], {
        ignoreInitial: true
      })
      .on('add', onChange)
      .on('change', onChange)
      .on('unlink', onChange)
  }

  // re-entrant method
  #watchBuild() {
    const localBuildId = ++this.#watch.buildId

    this.#watch.isBuilding = true
    this.#watch.rolldownWatcher = rolldownWatch(this.#getRolldownConfig()).on(
      'event',
      event => {
        if (localBuildId !== this.#watch.buildId) {
          event.result?.close?.()
          return
        }

        if (event.code === 'START') {
          const banner = ` ${basename(this.#ctx.appPaths.quasarConfigFilename)} (${this.#confEnv.envBanner})`
          this.#watch.isBuilding = true

          if (this.#watch.resolveRead === null) {
            log()
            log(`Recompiling${banner}...`)
          } else {
            log(`Compiling${banner}...`)
          }
        } else if (event.code === 'BUNDLE_END') {
          event.result.close()
          this.#watch.isBuilding = false

          if (this.#watch.resolveRead !== null) {
            this.#watch.resolveRead()
            this.#watch.resolveRead = null
            return
          }

          if (this.#watch.onUpdate === null) {
            this.#watch.shouldReadBuiltFileAgain = true
            return
          }

          this.#readBuiltFile().then(quasarConf => {
            if (quasarConf !== void 0 && localBuildId === this.#watch.buildId) {
              this.#injectAppEnv(quasarConf)
              this.#watch.onUpdate(quasarConf)
            }
          })
        } else if (event.code === 'ERROR') {
          fse.removeSync(this.#tempFile)

          const msg =
            'Could not compile the quasar.config file because it has errors.'

          if (this.#shouldFail) {
            error(msg, 'FAIL')
            console.error(event.error)
            process.exit(1)
          }

          warn(msg + ' Please fix them.\n')
          console.error(event.error)
          event.result.close()
        }
      }
    )
  }

  async #build() {
    log(
      `Compiling ${basename(this.#ctx.appPaths.quasarConfigFilename)} (${this.#confEnv.envBanner})`
    )

    let bundle
    try {
      const rolldownConfig = this.#getRolldownConfig()
      bundle = await rolldown(rolldownConfig)
      await bundle.write(rolldownConfig.output)
    } catch (err) {
      fse.removeSync(this.#tempFile)
      console.log()
      console.error(err)
      fatal(
        'Could not compile the quasar.config file because it has errors.',
        'FAIL'
      )
    }

    await bundle.close()
  }

  // exits process on first build, otherwise
  // returns undefined if it encounters errors
  async #computeConfig(quasarConfigFn) {
    if (typeof quasarConfigFn !== 'function') {
      fse.removeSync(this.#tempFile)

      const msg =
        'The default export value of the quasar.config file is not a function.'

      if (this.#shouldFail) fatal(msg, 'FAIL')

      warn(msg + ' Please fix it.\n')
      return
    }

    let userCfg

    try {
      userCfg = await quasarConfigFn(this.#ctx)
    } catch (err) {
      console.log()
      console.error(err)

      const msg =
        'The quasar.config file has runtime errors.' +
        ' Please check the Node.js stack above against the' +
        ` temporarily created ${basename(this.#tempFile)} file` +
        ' then DELETE it ("quasar clean --qconf" can be used).'

      if (this.#shouldFail) fatal(msg, 'FAIL')

      warn(msg + ' Please fix the errors in the original file.\n')
      return
    }

    fse.removeSync(this.#tempFile)

    if (Object(userCfg) !== userCfg) {
      const msg = 'The quasar.config file does not default exports an Object.'

      if (this.#shouldFail) fatal(msg, 'FAIL')

      warn(msg + ' Please fix it.\n')
      return
    }

    const { appPaths } = this.#ctx

    let rawQuasarConf = merge(
      {
        ctx: this.#ctx,

        boot: [],
        css: [],
        extras: [],
        animations: [],

        framework: {
          components: [],
          directives: [],
          plugins: [],
          config: {}
        },

        sourceFiles: {},
        bin: {},
        htmlVariables: {},

        devServer: {
          fs: {}
        },

        build: {
          target: {},
          viteVuePluginOptions: {},
          vitePlugins: [],
          define: {},
          defineEnv: {},
          env: {
            file: [],
            folder: [],
            ignoreType: []
          },
          alias: {},
          typescript: {},
          htmlMinifyOptions: {}
        },

        ssr: {
          middlewares: []
        },
        pwa: {},
        electron: {
          preloadScripts: [],
          unPackagedInstallParams: [],
          packager: {},
          builder: {}
        },
        cordova: {},
        capacitor: {
          capacitorCliPreparationParams: []
        },
        bex: {
          extraScripts: []
        }
      },
      userCfg
    )

    const metaConf = {
      debugging: Boolean(this.#ctx.dev || this.#ctx.debug),
      hasTypescript: await this.#ctx.cacheProxy.getModule('hasTypescript'),
      needsAppMountHook: false,
      vueDevtools: false,
      versions: { ...this.#versions }, // used by entry templates
      css: { ...this.#cssVariables }
    }

    if (rawQuasarConf.animations === 'all') {
      rawQuasarConf.animations =
        await this.#ctx.cacheProxy.getModule('animations')
    }

    try {
      await this.#ctx.appExt.runAppExtensionHook(
        'extendQuasarConf',
        async hook => {
          hook.api.logger.log('Extending quasar.config file configuration...')
          const overrides = await hook.fn(rawQuasarConf, hook.api)
          if (Object(overrides) === overrides) {
            rawQuasarConf = merge(rawQuasarConf, overrides)
          }
        }
      )
    } catch (err) {
      console.log()
      console.error(err)

      if (this.#shouldFail) {
        fatal('One of your installed App Extensions failed to run', 'FAIL')
      }

      warn('One of your installed App Extensions failed to run.\n')
      return
    }

    const cfg = {
      ...rawQuasarConf,
      metaConf
    }

    // we need to know if using SSR + PWA immediately
    if (this.#ctx.mode.ssr) {
      cfg.ssr = merge(
        {
          pwa: false,
          pwaOfflineHtmlFilename: 'offline.html',
          manualStoreHydration: false,
          manualPostHydrationTrigger: false,
          prodPort: 3000, // gets superseded in production by an eventual process.env.PORT
          prodScriptNamedExport: false
        },
        cfg.ssr
      )
    }

    if (this.#ctx.dev) {
      if (this.#opts.host) {
        cfg.devServer.host = this.#opts.host
      } else if (!cfg.devServer.host) {
        cfg.devServer.host = '0.0.0.0'
      }

      if (this.#opts.port) {
        cfg.devServer.port = this.#opts.port
        tip(
          'You are using the --port parameter. It is recommended to use a different devServer port for each Quasar mode to avoid browser cache issues'
        )
      } else if (!cfg.devServer.port) {
        cfg.devServer.port =
          defaultPortMapping[this.#ctx.modeName] +
          (this.#ctx.mode.ssr && cfg.ssr.pwa ? 50 : 0)
      } else {
        tip(
          'You (or an AE) specified an explicit quasar.config file > devServer > port. It is recommended to use' +
            ' a different devServer > port for each Quasar mode to avoid browser cache issues.' +
            ' Example: ctx.mode.ssr ? 9100 : ...'
        )
      }

      if (
        this.#address &&
        this.#address.from.host === cfg.devServer.host &&
        this.#address.from.port === cfg.devServer.port
      ) {
        cfg.devServer.host = this.#address.to.host
        cfg.devServer.port = this.#address.to.port
      } else {
        const addr = {
          host: cfg.devServer.host,
          port: cfg.devServer.port
        }
        const to = this.#opts.verifyAddress
          ? await onAddress(addr, this.#ctx.modeName)
          : addr

        // if network error while running
        if (to === null) {
          const msg =
            'Network error encountered while following the quasar.config file host/port config.'

          if (this.#shouldFail) fatal(msg, 'FAIL')

          warn(msg + ' Reconfigure and save the file again.\n')
          return
        }

        cfg.devServer = merge({ open: true }, cfg.devServer, to)
        this.#address = {
          from: addr,
          to: {
            host: cfg.devServer.host,
            port: cfg.devServer.port
          }
        }
      }
    }
    // else for production:
    else if (!Array.isArray(cfg.ssr.prodScriptNamedExport)) {
      cfg.ssr.prodScriptNamedExport = cfg.ssr.prodScriptNamedExport
        ? [cfg.ssr.prodScriptNamedExport]
        : []
    }

    if (cfg.css.length !== 0) {
      cfg.css = cfg.css
        .filter(Boolean)
        .map(parseAssetProperty('@/css'))
        .filter(asset => asset.path)
        .filter(uniquePathFilter)
    }

    if (cfg.boot.length !== 0) {
      cfg.boot = cfg.boot
        .filter(Boolean)
        .map(parseAssetProperty('@/boot'))
        .filter(asset => asset.path)
        .filter(uniquePathFilter)
    }

    if (cfg.extras.length !== 0) {
      cfg.extras = getUniqueArray(cfg.extras)
    }

    if (cfg.animations.length !== 0) {
      cfg.animations = getUniqueArray(cfg.animations)
    }

    if (
      !['kebab', 'pascal', 'combined'].includes(
        cfg.framework.autoImportComponentCase
      )
    ) {
      cfg.framework.autoImportComponentCase = 'kebab'
    }

    // special case where a component can be designated for a framework > config prop
    const { config } = cfg.framework

    if (config.loading) {
      const { spinner } = config.loading
      if (quasarComponentRE.test(spinner)) {
        cfg.framework.components.push(spinner)
      }
    }

    if (config.notify) {
      const { spinner } = config.notify
      if (quasarComponentRE.test(spinner)) {
        cfg.framework.components.push(spinner)
      }
    }

    cfg.framework.components = getUniqueArray(cfg.framework.components)
    cfg.framework.directives = getUniqueArray(cfg.framework.directives)
    cfg.framework.plugins = getUniqueArray(cfg.framework.plugins)

    const { lang, iconSet } = cfg.framework

    if (lang !== void 0) {
      cfg.framework.lang = formatQuasarAssetPath(lang, 'lang')
    }

    if (iconSet !== void 0) {
      cfg.framework.iconSet = formatQuasarAssetPath(iconSet, 'icon-set')
    }

    Object.assign(cfg.metaConf, {
      hasLoadingBarPlugin: cfg.framework.plugins.includes('LoadingBar'),
      hasMetaPlugin: cfg.framework.plugins.includes('Meta')
    })

    cfg.build = merge(
      {
        filenameBasedRouting: false,

        viteVuePluginOptions: {
          isProduction: this.#ctx.prod,
          template: {
            isProd: this.#ctx.prod,
            transformAssetUrls
          }
        },

        vueOptionsAPI: false,
        vueRouterMode: 'hash',

        minify:
          !cfg.metaConf.debugging &&
          Boolean(!this.#ctx.mode.bex || cfg.bex.minify),

        sourcemap: cfg.metaConf.debugging,

        useFilenameHashes: true,
        distDir: join('dist', this.#ctx.modeName),

        htmlMinifyOptions: {
          removeComments: true,
          collapseWhitespace: true,
          removeAttributeQuotes: true,
          collapseBooleanAttributes: true,
          removeScriptTypeAttributes: true
          // more options:
          // https://github.com/terser/html-minifier-terser?tab=readme-ov-file#options-quick-reference
        },

        define: {
          // vue
          __VUE_PROD_DEVTOOLS__: String(cfg.metaConf.debugging),
          __VUE_PROD_HYDRATION_MISMATCH_DETAILS__: String(
            cfg.metaConf.debugging
          ), // Vue 3.4+
          // __VUE_OPTIONS_API__ is set further down the line, based on cfg.build.vueOptionsAPI

          // vue-i18n
          __VUE_I18N_FULL_INSTALL__: 'true',
          __VUE_I18N_LEGACY_API__: 'false',
          __VUE_I18N_PROD_DEVTOOLS__: String(cfg.metaConf.debugging),
          __INTLIFY_PROD_DEVTOOLS__: String(cfg.metaConf.debugging)
        },

        env: {
          clientPrefix: defaultClientAppEnvPrefix,
          backendPrefix: defaultBackendAppEnvPrefix
        },

        typescript: {
          strict: false,
          vueShim: false
        }
      },
      cfg.build
    )

    // User-defined aliases take precedence over framework defaults.
    // We preserve the user insertion order for consumers (e.g. Vite, tsconfig `paths`)
    // Framework defaults come last if not already defined by the user.
    const { alias } = cfg.build
    const defaultAliases = {
      '@': appPaths.srcDir,
      '#q-app': '@quasar/app-vite'
    }
    for (const aliasKey in defaultAliases) {
      if (!(aliasKey in alias)) {
        alias[aliasKey] = defaultAliases[aliasKey]
      }
    }

    /**
     * If user hasn't explicitly set __VUE_OPTIONS_API__,
     * then we set it ourselves based on cfg.build.vueOptionsAPI.
     */
    if (cfg.build.define.__VUE_OPTIONS_API__ === void 0) {
      cfg.build.define.__VUE_OPTIONS_API__ = String(
        Boolean(cfg.build.vueOptionsAPI)
      )
    }

    Object.assign(
      cfg.build.define,
      Object.entries(cfg.build.defineEnv).reduce((acc, [key, value]) => {
        acc[`import.meta.env.${key}`] = JSON.stringify(value)
        return acc
      }, {})
    )
    /**
     * Now free up the memory taken by the original defineEnv object,
     * as it's not needed anymore
     */
    cfg.build.defineEnv = null

    /**
     * Avoid Rolldown failing to build with "undefined" value in define.
     *
     * Also, avoid mismatch between Vite & Rolldown on non
     * string values; Vite allows non stringified values, but
     * Rolldown doesn't and it fails the build.
     */
    const { define } = cfg.build
    for (const key in define) {
      if (define[key] === void 0) {
        delete define[key]
      } else if (typeof define[key] !== 'string') {
        define[key] = JSON.stringify(define[key])
      }
    }

    if (!cfg.build.target.browser) {
      cfg.build.target.browser = BASELINE_WIDELY_AVAILABLE_TARGET_STRING
    }

    if (!cfg.build.target.node) {
      cfg.build.target.node = 'node22'
    }

    if (cfg.build.filenameBasedRouting) {
      const defaultOptions = {
        // where are paths relative to:
        root: this.#ctx.appPaths.appDir,
        dts: cfg.metaConf.hasTypescript
          ? './src/router/typed-router.d.ts'
          : false
      }

      const { filenameBasedRouting } = cfg.build
      cfg.build.filenameBasedRouting =
        Object(filenameBasedRouting) === filenameBasedRouting
          ? merge({}, defaultOptions, filenameBasedRouting)
          : defaultOptions
    }

    if (this.#ctx.mode.ssr) {
      cfg.build.vueRouterMode = 'history'
    } else if (
      this.#ctx.mode.cordova ||
      this.#ctx.mode.capacitor ||
      this.#ctx.mode.electron ||
      this.#ctx.mode.bex
    ) {
      cfg.build.vueRouterMode = 'hash'
    }

    if (this.#ctx.mode.bex) {
      // we want to differentiate the folder
      // otherwise we can't run dev and build simultaneously;
      // it's better regardless because it's easier to select the dev folder
      // when loading the browser extension

      const name = basename(cfg.build.distDir)

      cfg.build.distDir = join(
        dirname(cfg.build.distDir),
        `bex-${this.#ctx.targetName}${name !== 'bex' ? `-${name}` : ''}${this.#ctx.dev ? '--dev' : ''}`
      )

      if (this.#ctx.dev) {
        // we need this token to remain the same across quasar.config changes + server restarts
        cfg.metaConf.bexWsToken = createWsToken()
      }
    }

    if (!isAbsolute(cfg.build.distDir)) {
      cfg.build.distDir = appPaths.resolve.app(cfg.build.distDir)
    }

    cfg.build.publicPath =
      cfg.build.publicPath && ['spa', 'pwa', 'ssr'].includes(this.#ctx.modeName)
        ? formatPublicPath(cfg.build.publicPath)
        : ['capacitor', 'cordova', 'electron', 'bex'].includes(
              this.#ctx.modeName
            )
          ? ''
          : '/'

    /* careful if you configure the following; make sure that you really know what you are doing */
    cfg.build.vueRouterBase =
      cfg.build.vueRouterBase !== void 0
        ? cfg.build.vueRouterBase
        : formatRouterBase(cfg.build.publicPath)

    // When adding new props here be sure to update
    // all impacted devserver diffs (look for this.registerDiff() calls)
    cfg.sourceFiles = merge(
      {
        // For Rolldown JS/TS entry points, make sure this
        // gets appPaths.resolve.app() applied to it further down the line

        rootComponent: 'src/App.vue',
        router: 'src/router/index',
        store: `src/${this.#storeProvider.pathKey}/index`,
        pwaRegisterServiceWorker: 'src-pwa/register-sw',
        pwaServiceWorker: 'src-pwa/sw/custom-sw',
        pwaManifestFile: 'src-pwa/manifest.json',
        electronMain: 'src-electron/electron-main',
        bexManifestFile: 'src-bex/manifest.json'
      },
      cfg.sourceFiles
    )

    if (!appFilesValidations(appPaths)) {
      if (this.#shouldFail) {
        fatal('Files validation not passed successfully', 'FAIL')
      }

      warn('Files validation not passed successfully. Please fix the issues.\n')
      return
    }

    // do we have a store?
    const storePath = appPaths.resolve.app(cfg.sourceFiles.store)
    Object.assign(cfg.metaConf, {
      hasStore: resolveExtension(storePath) !== void 0,
      storePackage: this.#storeProvider.name
    })

    // make sure we have preFetch in config
    cfg.preFetch ||= false

    if (
      this.#ctx.mode.capacitor &
      (cfg.capacitor.capacitorCliPreparationParams.length === 0)
    ) {
      cfg.capacitor.capacitorCliPreparationParams = [
        'sync',
        this.#ctx.targetName
      ]
    }

    if (this.#ctx.mode.ssr) {
      if (cfg.ssr.manualPostHydrationTrigger !== true) {
        cfg.metaConf.needsAppMountHook = true
      }

      cfg.ssr.middlewares =
        cfg.ssr.middlewares.length !== 0
          ? cfg.ssr.middlewares
              .filter(Boolean)
              .map(parseAssetProperty('@/../src-ssr/middlewares'))
              .filter(asset => asset.path)
              .filter(uniquePathFilter)
          : ['render'].map(parseAssetProperty('@/../src-ssr/middlewares'))

      if (cfg.ssr.pwa) {
        // install pwa mode if it's missing
        const { addMode } = await import('../lib/modes/pwa/pwa-installation.js')
        await addMode({ ctx: this.#ctx, silent: true })
      }

      this.#ctx.mode.pwa = cfg.ctx.mode.pwa = Boolean(cfg.ssr.pwa)
    }

    if (this.#ctx.dev) {
      if (this.#ctx.vueDevtools || cfg.devServer.vueDevtools) {
        if (this.#vueDevtools === void 0) {
          const host = localHostList.includes(cfg.devServer.host.toLowerCase())
            ? 'localhost'
            : cfg.devServer.host

          this.#vueDevtools = {
            host,
            port: await findClosestOpenPort(11_111, '0.0.0.0')
          }
        }

        cfg.metaConf.vueDevtools = { ...this.#vueDevtools }
      }

      if (this.#ctx.mode.electron || this.#ctx.mode.bex) {
        cfg.devServer.https = false
      } else if (
        cfg.devServer.open &&
        !this.#ctx.mode.cordova &&
        !this.#ctx.mode.capacitor
      ) {
        cfg.metaConf.openBrowser =
          !isCI && process.stdout.isTTY && process.env.NODE_ENV !== 'test'
            ? cfg.devServer.open
            : false
      }

      delete cfg.devServer.open
    }

    if (this.#ctx.dev) {
      if (cfg.devServer.https === true) {
        const { getCertificate } = await import('@quasar/ssl-certificate')
        const sslCertificate = await getCertificate({ log, fatal })
        cfg.devServer.https = {
          key: sslCertificate,
          cert: sslCertificate
        }
      } else if (Object(cfg.devServer.https) === cfg.devServer.https) {
        const { https } = cfg.devServer

        // we now check if config is specifying a file path
        // and we actually read the contents so we can later supply correct
        // params to the node HTTPS server
        ;['ca', 'pfx', 'key', 'cert'].forEach(prop => {
          if (typeof https[prop] === 'string') {
            try {
              https[prop] = readFileSync(https[prop])
            } catch (err) {
              console.error(err)
              console.log()
              delete https[prop]
              warn(
                `The devServer.https.${prop} file could not be read. Removed the config.`
              )
            }
          }
        })
      }
    }

    if (this.#ctx.mode.pwa) {
      cfg.pwa = merge(
        {
          workboxMode: 'GenerateSW',
          injectPWAMetaTags: true,
          swFilename: 'sw.js', // should be .js (as it's the distribution file, not the input file)
          manifestFilename: 'manifest.json',
          useCredentialsForManifestTag: false
        },
        cfg.pwa
      )

      if (!['GenerateSW', 'InjectManifest'].includes(cfg.pwa.workboxMode)) {
        const msg =
          `Workbox strategy "${cfg.pwa.workboxMode}" is invalid. ` +
          'Valid quasar.config file > pwa > workboxMode options are: GenerateSW or InjectManifest.'

        if (this.#shouldFail) fatal(msg, 'FAIL')

        warn(msg + ' Please fix it.\n')
        return
      }

      cfg.build.define['import.meta.env.QUASAR_SERVICE_WORKER_FILE'] =
        `"${cfg.build.publicPath}${cfg.pwa.swFilename}"`

      cfg.metaConf.pwaManifestFile = appPaths.resolve.app(
        cfg.sourceFiles.pwaManifestFile
      )

      cfg.sourceFiles.pwaServiceWorker = appPaths.resolve.app(
        cfg.sourceFiles.pwaServiceWorker
      )
    } else if (this.#ctx.mode.bex) {
      cfg.metaConf.bexManifestFile = appPaths.resolve.app(
        cfg.sourceFiles.bexManifestFile
      )
    }

    if (this.#ctx.dev) {
      const getUrl = hostname =>
        `http${cfg.devServer.https ? 's' : ''}://${hostname}:${cfg.devServer.port}${cfg.build.publicPath}`
      const hostname =
        cfg.devServer.host === '0.0.0.0' ? 'localhost' : cfg.devServer.host

      cfg.metaConf.APP_URL = this.#ctx.mode.bex
        ? 'index.html'
        : getUrl(hostname)
      cfg.metaConf.getUrl = getUrl
    } else if (
      this.#ctx.mode.cordova ||
      this.#ctx.mode.capacitor ||
      this.#ctx.mode.bex
    ) {
      cfg.metaConf.APP_URL = 'index.html'
    }

    Object.assign(cfg.build.define, {
      'import.meta.env.QUASAR_DEV': String(this.#ctx.dev),
      'import.meta.env.QUASAR_PROD': String(this.#ctx.prod),
      'import.meta.env.QUASAR_DEBUG': String(cfg.metaConf.debugging),

      'import.meta.env.QUASAR_MODE': `"${this.#ctx.modeName}"`,

      'import.meta.env.QUASAR_CLIENT': 'true',
      'import.meta.env.QUASAR_SERVER': 'false',

      'import.meta.env.QUASAR_VUE_ROUTER_MODE': `"${cfg.build.vueRouterMode}"`,
      'import.meta.env.QUASAR_VUE_ROUTER_BASE': `"${cfg.build.vueRouterBase}"`
    })

    quasarModesList.forEach(mode => {
      cfg.build.define[`import.meta.env.QUASAR_${mode.toUpperCase()}_MODE`] =
        String(Boolean(this.#ctx.mode[mode]))
    })

    if (
      this.#ctx.mode.bex ||
      this.#ctx.mode.capacitor ||
      this.#ctx.mode.cordova
    ) {
      cfg.build.define['import.meta.env.QUASAR_TARGET'] =
        `"${this.#ctx.targetName}"`
    }

    if (cfg.metaConf.APP_URL) {
      cfg.build.define['import.meta.env.QUASAR_APP_URL'] =
        `"${cfg.metaConf.APP_URL}"`
    }

    if (this.#ctx.mode.electron) {
      cfg.sourceFiles.electronMain = appPaths.resolve.app(
        cfg.sourceFiles.electronMain
      )

      if (!userCfg.electron?.preloadScripts) {
        cfg.electron.preloadScripts = ['electron-preload']
      }

      if (this.#ctx.dev) {
        if (this.#electronInspectPort === void 0) {
          this.#electronInspectPort = await findClosestOpenPort(
            userCfg.electron?.inspectPort || 5858,
            '127.0.0.1'
          )
        }

        cfg.electron.inspectPort = this.#electronInspectPort
      } else {
        const { ensureInstall, getDefaultName } =
          await this.#ctx.cacheProxy.getModule('electron')

        const icon = appPaths.resolve.electron('electron-assets/icons/icon.png')
        const builderIcon =
          process.platform === 'linux'
            ? // backward compatible (linux-512x512.png)
              existsSync(icon)
              ? icon
              : appPaths.resolve.electron(
                  'electron-assets/icons/linux-512x512.png'
                )
            : appPaths.resolve.electron('electron-assets/icons/icon')

        cfg.electron = merge(
          {
            packager: {
              asar: true,
              icon: appPaths.resolve.electron('electron-assets/icons/icon'),
              overwrite: true
            },
            builder: {
              appId: 'quasar-app',
              icon: builderIcon,
              productName:
                this.#ctx.pkg.appPkg.productName ||
                this.#ctx.pkg.appPkg.name ||
                'Quasar App',
              directories: {
                buildResources: appPaths.electronDir
              }
            }
          },
          cfg.electron,
          {
            packager: {
              dir: join(cfg.build.distDir, 'UnPackaged'),
              out: join(cfg.build.distDir, 'Packaged')
            },
            builder: {
              directories: {
                app: join(cfg.build.distDir, 'UnPackaged'),
                output: join(cfg.build.distDir, 'Packaged')
              }
            }
          }
        )

        if (cfg.ctx.bundlerName) {
          cfg.electron.bundler = cfg.ctx.bundlerName
        } else if (!cfg.electron.bundler) {
          cfg.electron.bundler = getDefaultName()
        }

        ensureElectronArgv(cfg.electron.bundler, this.#ctx)

        if (cfg.electron.bundler === 'packager') {
          if (cfg.ctx.targetName) {
            cfg.electron.packager.platform = cfg.ctx.targetName
          }
          if (cfg.ctx.archName) {
            cfg.electron.packager.arch = cfg.ctx.archName
          }
        } else {
          cfg.electron.builder = {
            config: cfg.electron.builder,
            /**
             * Needed to determine the Electron version:
             */
            projectDir: appPaths.electronDir
          }

          if (
            cfg.ctx.targetName === 'mac' ||
            cfg.ctx.targetName === 'darwin' ||
            cfg.ctx.targetName === 'all'
          ) {
            cfg.electron.builder.mac = []
          }

          if (cfg.ctx.targetName === 'linux' || cfg.ctx.targetName === 'all') {
            cfg.electron.builder.linux = []
          }

          if (
            cfg.ctx.targetName === 'win' ||
            cfg.ctx.targetName === 'win32' ||
            cfg.ctx.targetName === 'all'
          ) {
            cfg.electron.builder.win = []
          }

          if (cfg.ctx.archName) {
            cfg.electron.builder[cfg.ctx.archName] = true
          }

          if (cfg.ctx.publish) {
            cfg.electron.builder.publish = cfg.ctx.publish
          }
        }

        await ensureInstall(cfg.electron.bundler)
      }
    }

    const entryScriptAbsolutePath = appPaths.resolve.entry('client-entry.js')
    const entryScriptWebPath = relative(
      appPaths.appDir,
      entryScriptAbsolutePath
    )

    cfg.metaConf.entryScript = {
      absolutePath: entryScriptAbsolutePath,
      webPath: cfg.build.publicPath + entryScriptWebPath,
      // publicPath will be handled by Vite middleware:
      tag: `<script type="module" src="/${entryScriptWebPath}"></script>`
    }
    cfg.htmlVariables = merge(
      {
        productName: escapeHTMLTagContent(this.#ctx.pkg.appPkg.productName),
        productDescription: escapeHTMLAttribute(
          this.#ctx.pkg.appPkg.description
        )
      },
      cfg.htmlVariables,
      /**
       * We protect the ctx from user modifications
       */
      { ctx: cfg.ctx }
    )

    if (
      this.#ctx.mode.capacitor &&
      cfg.metaConf.versions.capacitorPluginSplashscreen &&
      cfg.capacitor.hideSplashscreen !== false
    ) {
      cfg.metaConf.needsAppMountHook = true
    }

    return cfg
  }
}
