// oxfmt/dist/apis-DLO0lBDu.js
import { o as __toESM } from "./rolldown-runtime-CE-6LUnI.js";
//#region src-js/libs/apis.ts
const CACHES = {
	prettier: null,
	sveltePlugin: null,
	tailwindPlugin: null,
	tailwindSorter: null,
	oxfmtPlugin: null
};
async function loadCached(key, loader) {
	CACHES[key] ??= await loader();
	return CACHES[key];
}
async function loadPrettier() {
	return loadCached("prettier", async () => {
		const prettier = await import("./prettier-BByt4AhK.js");
		const { formatOptionsHiddenDefaults } = prettier.__internal;
		formatOptionsHiddenDefaults.parentParser = null;
		formatOptionsHiddenDefaults.__onHtmlRoot = null;
		formatOptionsHiddenDefaults.__inJsTemplate = null;
		return prettier;
	});
}
/**
* Format non-js file
*
* @returns Formatted code
*/
async function formatFile({ code, options }) {
	const prettier = CACHES.prettier ?? await loadPrettier();
	if ("_useSveltePlugin" in options) await setupSveltePlugin(options);
	if ("_useTailwindPlugin" in options) await setupTailwindPlugin(options);
	if ("_oxfmtPluginOptionsJson" in options) await setupOxfmtPlugin(options);
	return prettier.format(code, options);
}
/**
* Format non-js code snippets into formatted string.
* Mainly used for formatting code fences within JSDoc,
* and is also used as a temporary fallback for html-in-js.
*
* @returns Formatted code snippet
*/
async function formatEmbeddedCode({ code, options }) {
	const prettier = CACHES.prettier ?? await loadPrettier();
	if ("_useTailwindPlugin" in options) await setupTailwindPlugin(options);
	return prettier.format(code, options);
}
/**
* Format non-js code snippets into Prettier `Doc` JSON strings.
*
* This makes our printer correctly handle `printWidth` even for embedded code.
* - For gql-in-js, `texts` contains multiple parts split by `${}` in a template literal
* - For others, `texts` always contains a single string with `${}` parts replaced by placeholders
* However, this function does not need to be aware of that,
* as it simply formats each text part independently and returns an array of formatted parts.
*
* @returns Doc JSON strings
*/
async function formatEmbeddedDoc({ texts, options }) {
	const prettier = CACHES.prettier ?? await loadPrettier();
	if ("_useTailwindPlugin" in options) await setupTailwindPlugin(options);
	return Promise.all(texts.map(async (text) => {
		const metadata = {};
		if (options.parser === "html" || options.parser === "angular") {
			options.parentParser = "OXFMT";
			options.__onHtmlRoot = (root) => metadata.htmlHasMultipleRootElements = (root.children?.length ?? 0) > 1;
		}
		if (options.parser === "markdown") options.__inJsTemplate = true;
		const doc = await prettier.__debug.printToDoc(text, options);
		const symbolToNumber = /* @__PURE__ */ new Map();
		let nextId = 1;
		return JSON.stringify([doc, metadata], (_key, value) => {
			if (typeof value === "symbol") {
				if (!symbolToNumber.has(value)) symbolToNumber.set(value, nextId++);
				return symbolToNumber.get(value);
			}
			if (value === -Infinity) return "__NEGATIVE_INFINITY__";
			return value;
		});
	}));
}
/**
* Load Tailwind CSS plugin.
* Option mapping (sortTailwindcss.xxx → tailwindXxx) is also done in Rust side.
*/
async function setupTailwindPlugin(options) {
	CACHES.tailwindPlugin ??= await loadCached("tailwindPlugin", () => import("./dist-YOS4M6hk.js"));
	options.plugins ??= [];
	options.plugins.push(CACHES.tailwindPlugin);
}
/**
* Process Tailwind CSS classes found in JS/TS files in batch.
* @param args - Object containing classes and options (filepath is in options.filepath)
* @returns Array of sorted class strings (same order/length as input)
*/
async function sortTailwindClasses({ classes, options }) {
	CACHES.tailwindSorter ??= await loadCached("tailwindSorter", () => import("./sorter-BcGkkjyR.js"));
	const { createSorter } = CACHES.tailwindSorter;
	return (await createSorter({
		filepath: options.filepath,
		stylesheetPath: options.tailwindStylesheet,
		configPath: options.tailwindConfig,
		preserveWhitespace: options.tailwindPreserveWhitespace,
		preserveDuplicates: options.tailwindPreserveDuplicates
	})).sortClassAttributes(classes);
}
/**
* Load prettier-plugin-svelte to provide the `svelte` parser.
*/
async function setupSveltePlugin(options) {
	CACHES.sveltePlugin ??= await loadCached("sveltePlugin", async () => await import("./plugin-B4cguZ1l.js").then((m) => /* @__PURE__ */ __toESM(m.default, 1)));
	options.plugins ??= [];
	options.plugins.push(CACHES.sveltePlugin);
}
/**
* Load oxfmt plugin for js-in-xxx parsers.
*/
async function setupOxfmtPlugin(options) {
	CACHES.oxfmtPlugin ??= await loadCached("oxfmtPlugin", async () => await import("./prettier-plugin-oxfmt-Dj6FaAil.js"));
	options.plugins ??= [];
	options.plugins.push(CACHES.oxfmtPlugin);
}
//#endregion
export { sortTailwindClasses as i, formatEmbeddedDoc as n, formatFile as r, formatEmbeddedCode as t };
